### PR TITLE
Bug 1822447 - Upgrade RecyclerView Library and remove tab tray thumbnail placeholder

### PR DESCRIPTION
**Pull request by HarrisonOg**
### The actual problem being solved:
After the tabs tray is opened, the thumbnails are painted to the screen much later.  This results in a visual "flash" loading which is an undesirable experience for users.

### Learnings:
- There are a large amount of Compose-related processes that are delaying the painting of the thumbnails after the thumbnails are requested from storage.  These Compose-related processes are from the Inactive Tabs section of the tabs tray.
- I confirmed the processes were connected to Inactive Tabs by removing the Inactive Tabs section and the tab thumbnails were able to load before the tabs tray was fully extended.
- The Inactive Tabs section cannot be removed from the tabs tray for production so performance enhancements were the only thing we can do before we convert the entire Tabs Tray to Compose.

### The solution in this PR:
- Update `androidx.recyclerview:recyclerview` to `1.3.0`
- Remove placeholder "globe" image from tab tray items to prevent the image from loading before the thumbnail loads.

After landing this, we will re-evaluate the problem after the Tabs Tray to Compose project concludes.  If the problem persists, we will look into more solutions after that.

Link to Bugzilla comment on why the upgrade to new version of `RecyclerView` : https://bugzilla.mozilla.org/show_bug.cgi?id=1822447#c17
Link to developer notes on the `1.3.0` update to RecyclerView: https://developer.android.com/jetpack/androidx/releases/recyclerview#recyclerview-1.3.0

This PR currently will only apply the changes for 1.3.0 to the RecyclerView in the Tabs Tray with a followup for other RecyclerViews coming in a separate PR.  [Followup PR](https://github.com/mozilla-mobile/firefox-android/pull/1878)

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1822447

### CODE REVIEW

File: fenix/app/src/main/res/layout/tab_tray_grid_item.xml [hunk 1]
```diff
@@ -102,7 +102,6 @@ A FrameLayout here is an efficient way of having a views stack while allowing:
                     android:layout_height="match_parent"
                     android:importantForAccessibility="no"
                     android:padding="22dp"
-                    app:srcCompat="@drawable/mozac_ic_globe"
```
Comment (MozillaNoah) [inline fenix/app/src/main/res/layout/tab_tray_grid_item.xml:105]: should the entire `ImageView` just be removed?

The icon in the sister list_item XML will need to be cleaned-up, too.

Comment (HarrisonOg) [reply fenix/app/src/main/res/layout/tab_tray_grid_item.xml:105]: We need the `TabThumbnailView` still to attach the thumbnail but I can get rid of the unnecessary `ImageView` with the id `default_tab_thumbnail`
on both the grid item and `tab_tray_item`

Comment (jonalmeida) [reply fenix/app/src/main/res/layout/tab_tray_grid_item.xml:105]: Shouldn't we ask UX before we remove the globe placeholder? I don't see any confirmation about this on the bug besides comment 6 from verdi, and this was before we learnt that the cause was Compose inside a RecyclerView.

If there was confirmation, please link it in the PR and add remember to add the label for UX to review before we merge.

**Comment by jonalmeida** [comment]
Please add a description into the PR to explain why we are doing what we are doing. Reasoning: we are updating the RecyclerView library for _all_ lists in the app for a perf fix in only one area (currently). We should be more clear in PRs like this what is the magnitude of it.

At the very least, link to the comment in the bug (comment 17) where you made your original analysis.

**Comment by jonalmeida** [comment]
> [task 2023-04-20T21:33:40.697Z] * What went wrong:
> [task 2023-04-20T21:33:40.697Z] Execution failed for task ':app:dataBindingMergeDependencyArtifactsFenixDebug'.
> [task 2023-04-20T21:33:40.697Z] > Could not resolve all files for configuration ':app:fenixDebugCompileClasspath'.
> [task 2023-04-20T21:33:40.697Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
> [task 2023-04-20T21:33:40.697Z]      Required by:
> [task 2023-04-20T21:33:40.697Z]          project :app
> [task 2023-04-20T21:33:40.697Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
> [task 2023-04-20T21:33:40.697Z]      Required by:
> [task 2023-04-20T21:33:40.697Z]          project :app > project :browser-tabstray
> [task 2023-04-20T21:33:40.697Z]          project :app > androidx.paging:paging-runtime:3.1.1
> [task 2023-04-20T21:33:40.697Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
> [task 2023-04-20T21:33:40.697Z]      Required by:
> [task 2023-04-20T21:33:40.697Z]          project :app > com.google.android.material:material:1.4.0-beta01
> [task 2023-04-20T21:33:40.697Z]          project :app > androidx.preference:preference:1.1.1
> [task 2023-04-20T21:33:40.697Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
> [task 2023-04-20T21:33:40.697Z]      Required by:
> [task 2023-04-20T21:33:40.697Z]          project :app > androidx.viewpager2:viewpager2:1.0.0

This failure seems like a bug in our CI. I suspect we need to add `fenix/buildSrc/src/main/java/FenixDependencies.kt` to [this list of resources](https://searchfox.org/mozilla-mobile/rev/8a11e34d36710a20e0ba4653dd548ba0c6061419/firefox-android/taskcluster/ci/external-gradle-dependencies/kind.yml#27) that re-trigger a fetch of our external gradle dependencies.

@JohanLorenzo does that sound right? If so, I can file a bug and we can touch the AC dependency list for this PR to workaround the issue.

**Comment by rvandermeulen** [comment]
> I suspect we need to add fenix/buildSrc/src/main/java/FenixDependencies.kt to [this list of resources](https://searchfox.org/mozilla-mobile/rev/8a11e34d36710a20e0ba4653dd548ba0c6061419/firefox-android/taskcluster/ci/external-gradle-dependencies/kind.yml#27) that re-trigger a fetch of our external gradle dependencies.

That seems plausible to me, fwiw. Note that we probably also need to do the same for Focus if that's the case.

File: fenix/app/src/main/res/layout/tab_tray_grid_item.xml (hunk 1 above)
Comment (HarrisonOg) [reply fenix/app/src/main/res/layout/tab_tray_grid_item.xml:105]: I'll separate out the globe image removal to a separate PR so we can go ahead with this upgrade (and the following PR to apply the upgrades to the other RecyclerViews in the app).

**Comment by HarrisonOg** [comment]
re-removed the globe image from the tab tray items as this was also approved by Design.

**Comment by jonalmeida** [comment]
> > I suspect we need to add fenix/buildSrc/src/main/java/FenixDependencies.kt to [this list of resources](https://searchfox.org/mozilla-mobile/rev/8a11e34d36710a20e0ba4653dd548ba0c6061419/firefox-android/taskcluster/ci/external-gradle-dependencies/kind.yml#27) that re-trigger a fetch of our external gradle dependencies.
> 
> That seems plausible to me, fwiw. Note that we probably also need to do the same for Focus if that's the case.

Thanks for the catch! I've messaged @HarrisonOg offline that we should do this as part of their follow-up in https://github.com/mozilla-mobile/firefox-android/pull/1878.

What we're seeing in this PR is that the 1.3.0 is upgrading AC implicitly, so we should make that _explicit_ soon and update Focus too.

Screenshot 2023-05-03 at 2 03 11 PM

**Comment by jonalmeida** [comment]
build-apk-fenix-debug  is still failing because of a "missing" dependency, so the cache is still wrong.

```
[task 2023-05-03T16:49:40.422Z] * What went wrong:
[task 2023-05-03T16:49:40.422Z] Execution failed for task ':app:dataBindingMergeDependencyArtifactsFenixDebug'.
[task 2023-05-03T16:49:40.422Z] > Could not resolve all files for configuration ':app:fenixDebugCompileClasspath'.
[task 2023-05-03T16:49:40.422Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
[task 2023-05-03T16:49:40.422Z]      Required by:
[task 2023-05-03T16:49:40.422Z]          project :app
[task 2023-05-03T16:49:40.422Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
[task 2023-05-03T16:49:40.422Z]      Required by:
[task 2023-05-03T16:49:40.422Z]          project :app > project :browser-tabstray
[task 2023-05-03T16:49:40.422Z]          project :app > androidx.paging:paging-runtime:3.1.1
[task 2023-05-03T16:49:40.422Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
[task 2023-05-03T16:49:40.422Z]      Required by:
[task 2023-05-03T16:49:40.422Z]          project :app > com.google.android.material:material:1.4.0-beta01
[task 2023-05-03T16:49:40.422Z]          project :app > androidx.preference:preference:1.1.1
[task 2023-05-03T16:49:40.422Z]    > Could not find androidx.recyclerview:recyclerview:1.3.0.
[task 2023-05-03T16:49:40.422Z]      Required by:
[task 2023-05-03T16:49:40.422Z]          project :app > androidx.viewpager2:viewpager2:1.0.0
```

@HarrisonOg can you touch [this file](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt#L9) so that the cache thinks there is a change and retriggers the AC cache?

**Review by jonalmeida — APPROVED** [review]
Looks good to me! Thanks for pushing this through and doing all the hardwork! 💯 

If CI doesn't complain, let's get this landed before the cut.

**Comment by AaronMT** [comment]
Hey @HarrisonOg, I see a consistent 2 failures on `ui-test-apk-fenix-arm-debug` on this PR

```Kotlin
androidx.test.uiautomator.UiObjectNotFoundException: UiSelector[RESOURCE_ID=org.mozilla.fenix.debug:id/privateBrowsingButton]
	at androidx.test.uiautomator.UiObject.click(UiObject.java:416)
	at org.mozilla.fenix.ui.robots.HomeScreenRobot$Transition.togglePrivateBrowsingMode(HomeScreenRobot.kt:604)
	at org.mozilla.fenix.ui.SmokeTest.privateTabsTrayWithOpenedTabTest(SmokeTest.kt:298)
```

and in tab thumbnail checks

```java.lang.AssertionError
at org.junit.Assert.fail(Assert.java:87)
at org.junit.Assert.assertTrue(Assert.java:42)
at org.junit.Assert.assertTrue(Assert.java:53)
at org.mozilla.fenix.helpers.MatcherHelper.assertItemWithResIdExists(MatcherHelper.kt:54)
at org.mozilla.fenix.helpers.MatcherHelper.assertItemWithResIdExists$default(MatcherHelper.kt:51)
at org.mozilla.fenix.ui.robots.TabDrawerRobot.verifyOpenedTabThumbnail(TabDrawerRobot.kt:106)
at org.mozilla.fenix.ui.TabbedBrowsingTest$verifyOpenTabDetails$3.invoke(TabbedBrowsingTest.kt:328)
at org.mozilla.fenix.ui.TabbedBrowsingTest$verifyOpenTabDetails$3.invoke(TabbedBrowsingTest.kt:320)
at org.mozilla.fenix.ui.robots.BrowserRobot$Transition.openTabDrawer(BrowserRobot.kt:869)
at org.mozilla.fenix.ui.TabbedBrowsingTest.verifyOpenTabDetails(TabbedBrowsingTest.kt:320)
```

File: fenix/buildSrc/src/main/java/FenixDependencies.kt [hunk 2]
```diff
@@ -3,7 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 // If you ever need to force a toolchain rebuild (taskcluster) then edit the following comment.
-// FORCE REBUILD 2023-05-04
+// FORCE REBUILD 2023-05-05
```
Comment (rvandermeulen) [inline fenix/buildSrc/src/main/java/FenixDependencies.kt:6]: FWIW, this isn't needed if you're making actual changes to the dependency list anyway. Same for the AC one.

Comment (HarrisonOg) [reply fenix/buildSrc/src/main/java/FenixDependencies.kt:6]: gotcha, I'll remember for next time

**Pull request merged**